src/raw/mod.rs

```diff
@@ -274,42 +268,20 @@ pub type CompiledAddr = usize;
 ///   (excellent for in depth overview)
 /// * [Comparison of Construction Algorithms for Minimal, Acyclic, Deterministic, Finite-State Automata from Sets of Strings](http://www.cs.mun.ca/~harold/Courses/Old/CS4750/Diary/q3p2qx4lv71m5vew.pdf)
 ///   (excellent for surface level overview)
-pub struct Fst {
+pub struct Fst<D: AsRef<[u8]>> {
+    meta: Meta,
+    data: D,
+}
```

Comment:
[Continuing the discussion](https://github.com/BurntSushi/fst/commit/e33e730e3cd36210adf62241c3cddcbe9f3d85f9#r37473048) I miss-started on a commit instead of the PR itself.

-----

If a user wants to write a structure that stores a `fst::Map<D>` he is forced to specify that the D type is `AsRef<[u8]>` even if he will not be read or used but only stored.

```rust
struct CurrentStruct<D: AsRef<[u8]>> {
    fst: fst::Map<D>,
}

struct ExpectedStruct<D> {
    fst: fst::Map<D>,
}
```

Reply:
Sure, I understand that as a technical restriction. But _why_ would someone want to do that? Is there a specific use case you have in mind?

Reply:
> But why would someone want to do that?

If you mean: Why would someone want to wrap/store an FST inside of a struct that do not need to read the FST itself?

I think about one little thing that would not be possible, for example a generic struct that just displays the `size_of` a wrapped type, or displays the [`type name`](https://doc.rust-lang.org/std/any/fn.type_name.html) of a struct.

Don't know if it make any sense but common [std](https://doc.rust-lang.org/std/collections/struct.HashMap.html) [collections](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) do not trait constrain at the struct level, that makes users lives easier, even if the generic types must impl a bunch of traits to make the collections useful (`K: Hash`, or `K: Ord`).

Reply:
I'll consider it, thanks. And yes, I know that the std collections don't do it, but it's not clear to me whether that was a mistake or not.

Reply:
There is many arguments which defends my point in [this recent issue](https://github.com/rust-lang/rust/issues/44777) in the Rust repo. [A PR has been merged](https://github.com/rust-lang/rust/pull/67642) to relax those bounds. I think that it is important to be the less restrictive possible when unneeded, but that's only my point of view.

Reply:
Right, and in that PR, my cohorts on the library team brought up concerns with doing it. That's basically my stance as well. I just really haven't seen a compelling use case. Like I said, I'll consider it.

Reply:
I just wanted to see if I was irrelevant so I checked the API Guideline, [the Exception point](https://rust-lang.github.io/api-guidelines/future-proofing.html#exceptions) is on my side. I don't want to fight, I am just trying to make the library more flexible :)

Reply:
Not really. That guideline is appropriately circumspect:

> There is a grey area around other non-derivable trait bounds that are not strictly required by the structure definition, like Read or Write. They may communicate the **intended behavior of the type better in its definition** but also limits future extensibility. Including **semantically useful trait bounds** on data structures is still less problematic than including derivable traits as bounds.

We are spending too much time discussing a very minor point. _Like I said_, I'll consider it.